From f9b73316e9582832f0f793d2c66af0792f82b6bf Mon Sep 17 00:00:00 2001
From: He Junyan <junyan.he@intel.com>
Date: Mon, 12 Sep 2022 14:38:48 +0800
Subject: [PATCH 11/13] va: encoder: Add the P list 0 num into
 max_num_reference query.

The driver may report different reference frame number for P and B
frames' list 0 now. For example, the HW limitation of reference
number for encoding each frame is 3. When the frame is P, all the
3 references can be assigned to list 0 and no one in list 1. But
for B frame, 2 should be assigned to list 0 and 1 should be assigned
to list 1.
---
 .../gst-plugins-bad/sys/va/gstvaencoder.c     | 31 +++++++++++++------
 .../gst-plugins-bad/sys/va/gstvaencoder.h     |  5 +--
 .../gst-plugins-bad/sys/va/gstvah264enc.c     |  2 +-
 .../gst-plugins-bad/sys/va/gstvah265enc.c     |  2 +-
 .../gst-plugins-bad/sys/va/gstvavp9enc.c      |  2 +-
 5 files changed, 28 insertions(+), 14 deletions(-)

diff --git a/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c b/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c
index 9e08df713e..a715cc2799 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c
@@ -635,7 +635,7 @@ gst_va_encoder_get_slice_structure (GstVaEncoder * self,
 gboolean
 gst_va_encoder_get_max_num_reference (GstVaEncoder * self,
     VAProfile profile, VAEntrypoint entrypoint,
-    guint32 * list0, guint32 * list1)
+    guint32 * listp_0, guint32 * listb_0, guint32 * listb_1)
 {
   VAStatus status;
   VADisplay dpy;
@@ -655,18 +655,31 @@ gst_va_encoder_get_max_num_reference (GstVaEncoder * self,
   }
 
   if (attrib.value == VA_ATTRIB_NOT_SUPPORTED) {
-    if (list0)
-      *list0 = 0;
-    if (list1)
-      *list1 = 0;
+    if (listp_0)
+      *listp_0 = 0;
+    if (listb_0)
+      *listb_0 = 0;
+    if (listb_1)
+      *listb_1 = 0;
 
     return TRUE;
   }
 
-  if (list0)
-    *list0 = attrib.value & 0xffff;
-  if (list1)
-    *list1 = (attrib.value >> 16) & 0xffff;
+  if (profile == VAProfileAV1Profile0 || profile == VAProfileAV1Profile1) {
+    if (listp_0)
+      *listp_0 = attrib.value & 0xff;
+    if (listb_0)
+      *listb_0 = (attrib.value >> 8) & 0xff;
+    if (listb_1)
+      *listb_1 = (attrib.value >> 16) & 0xff;
+  } else {
+    if (listb_0)
+      *listb_0 = attrib.value & 0xffff;
+    if (listb_1)
+      *listb_1 = (attrib.value >> 16) & 0xffff;
+    if (listp_0)
+      *listp_0 = attrib.value & 0xffff;
+  }
 
   return TRUE;
 }
diff --git a/subprojects/gst-plugins-bad/sys/va/gstvaencoder.h b/subprojects/gst-plugins-bad/sys/va/gstvaencoder.h
index 290fb3531a..d233693be1 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvaencoder.h
+++ b/subprojects/gst-plugins-bad/sys/va/gstvaencoder.h
@@ -70,8 +70,9 @@ gint32                gst_va_encoder_get_slice_structure  (GstVaEncoder * self,
 gboolean              gst_va_encoder_get_max_num_reference (GstVaEncoder * self,
                                                             VAProfile profile,
                                                             VAEntrypoint entrypoint,
-                                                            guint32 * list0,
-                                                            guint32 * list1);
+                                                            guint32 * listp_0,
+                                                            guint32 * listb_0,
+                                                            guint32 * listb_1);
 guint                 gst_va_encoder_get_prediction_direction (GstVaEncoder * self,
                                                                VAProfile profile,
                                                                VAEntrypoint entrypoint);
diff --git a/subprojects/gst-plugins-bad/sys/va/gstvah264enc.c b/subprojects/gst-plugins-bad/sys/va/gstvah264enc.c
index be6a48f136..9df1e21f41 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvah264enc.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvah264enc.c
@@ -983,7 +983,7 @@ _generate_gop_structure (GstVaH264Enc * self)
   }
 
   if (!gst_va_encoder_get_max_num_reference (base->encoder, base->profile,
-          GST_VA_BASE_ENC_ENTRYPOINT (base), &list0, &list1)) {
+          GST_VA_BASE_ENC_ENTRYPOINT (base), NULL, &list0, &list1)) {
     GST_INFO_OBJECT (self, "Failed to get the max num reference");
     list0 = 1;
     list1 = 0;
diff --git a/subprojects/gst-plugins-bad/sys/va/gstvah265enc.c b/subprojects/gst-plugins-bad/sys/va/gstvah265enc.c
index 8e031510ea..0c4ef7798e 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvah265enc.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvah265enc.c
@@ -3803,7 +3803,7 @@ _h265_generate_gop_structure (GstVaH265Enc * self)
   }
 
   if (!gst_va_encoder_get_max_num_reference (base->encoder, base->profile,
-          GST_VA_BASE_ENC_ENTRYPOINT (base), &list0, &list1)) {
+          GST_VA_BASE_ENC_ENTRYPOINT (base), NULL, &list0, &list1)) {
     GST_INFO_OBJECT (self, "Failed to get the max num reference");
     list0 = 1;
     list1 = 0;
diff --git a/subprojects/gst-plugins-bad/sys/va/gstvavp9enc.c b/subprojects/gst-plugins-bad/sys/va/gstvavp9enc.c
index 4c8ec74dc0..cda716d229 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvavp9enc.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvavp9enc.c
@@ -1629,7 +1629,7 @@ _vp9_generate_gop_structure (GstVaVp9Enc * self)
 
   /* VP9 does not define reference list1 in spec. */
   if (!gst_va_encoder_get_max_num_reference (base->encoder, base->profile,
-          GST_VA_BASE_ENC_ENTRYPOINT (base), &list0, NULL)) {
+          GST_VA_BASE_ENC_ENTRYPOINT (base), NULL, &list0, NULL)) {
     GST_INFO_OBJECT (self, "Failed to get the max num reference");
     list0 = 1;
   }
-- 
2.25.1

